kernel/ps2/src/lib.rs

```diff
@@ -1,484 +1,704 @@
-pub fn data_to_port1(value: u8) -> u8 {
-    unsafe { PS2_PORT.lock().write(value) };
-    let response = ps2_read_data();
-    response
+#[derive(Clone)]
+enum HostToDevice {
+    Keyboard(HostToKeyboardCommandOrData),
+    Mouse(HostToMouseCommandOrData)
 }
 
-/// write data to the second ps2 data port and return the response
-pub fn data_to_port2(value: u8) -> u8 {
-    ps2_write_command(0xD4);
-    data_to_port1(value)
+// https://wiki.osdev.org/PS/2_Keyboard#Commands
+#[derive(Clone)]
+pub enum HostToKeyboardCommandOrData {
+    KeyboardCommand(HostToKeyboardCommand),
+    LEDState(LEDState),
+    ScancodeSet(ScancodeSet),
+    //TODO: Typematic, Scancode
 }
 
-/// write command to the keyboard and return the result
-pub fn command_to_keyboard(value: u8) -> Result<(), &'static str> {
-    let response = data_to_port1(value);
+#[derive(Clone)]
+pub enum HostToKeyboardCommand {
+    SetLEDStatus = 0xED,
+    ///// for diagnostic purposes and useful for device removal detection
+    //Echo = 0xEE,
+    SetScancodeSet = 0xF0,
+    IdentifyKeyboard = 0xF2,
+    /// also called typematic
+    SetRepeatRateAndDelay = 0xF3,
+    EnableScanning = 0xF4,
+    /// might also restore default parameters
+    DisableScanning = 0xF5,
+    //SetDefaultParameters = 0xF6,
+    ///// scancode set 3 only
+    //SetAllKeysToAutorepeat = 0xF7,
+    ///// scancode set 3 only
+    //SetAllKeysToMakeRelease = 0xF8,
+    ///// scancode set 3 only
+    //SetAllKeysToMake = 0xF9,
+    ///// scancode set 3 only
+    //SetAllKeysToAutorepeatMakeRelease = 0xFA, //NOTE: same value as ACK
+    ///// scancode set 3 only
+    //SetKeyToAutorepeat = 0xFB,
+    ///// scancode set 3 only
+    //SetKeyToMakeRelease = 0xFC, //NOTE: same value as Self test failed
+    ///// scancode set 3 only
+    //SetKeyToMake = 0xFD,
+    ResendByte = 0xFE,
+    ResetAndStartSelfTest = 0xFF,
+}
 
-    match response {
-        0xFA => {
-            // keyboard acknowledges the command
-            Ok(())
-        }
+#[bitfield(bits = 3)]
+#[derive(Clone)]
+pub struct LEDState {
+    pub scroll_lock: bool,
+    pub number_lock: bool,
+    pub caps_lock: bool,
+}
 
-        0xFE => {
-            // keyboard doesn't acknowledge the command
-            Err("Fail to send the command to the keyboard")
-        }
+#[derive(Clone)]
+pub enum ScancodeSet {
+    GetCurrentSet = 0,
+    Set1 = 1,
+    Set2 = 2,
+    Set3 = 3,
+}
 
-        _ => {
-            for _x in 0..14 {
-                // wait for response
-                let response = ps2_read_data();
-                if response == 0xFA {
-                    return Ok(());
-                } else if response == 0xfe {
-                    return Err("Please resend the command to the keyboard");
-                }
-            }
-            Err("command is not acknowledged")
+/// write command to the keyboard and return the result
+fn command_to_keyboard(value: HostToKeyboardCommandOrData) -> Result<(), &'static str> {
+    for _ in 0..3 {
```

Comment:
why do we try 3 times? i assume this is similar to the hilarious "VERY_ARBITRARY..." thing above?